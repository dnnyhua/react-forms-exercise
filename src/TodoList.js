import { useState } from 'react';
import NewTodoForm from './NewTodoForm'
import Todo from './Todo';
import { v4 as uuid } from 'uuid';

const TodoList = () => {
    const INITIAL_STATE = [
        { todoTask: "Walk the dog" }
    ]

    const [todos, setTodos] = useState(INITIAL_STATE);
    // a parent function that will be passed down as a prop to the boxForm so that data can be passed up to update the state and re-render
    const addTodo = (newTodo) => {
        setTodos(todos => [...todos, { ...newTodo, id: uuid() }])
    }
    // this will be passed down as a prop to the Todo component
    const removeTodo = (id) => {
        setTodos(todos => todos.filter(todo => todo.id !== id))
    }

    const update = (id, updatedTodoTask) => {
        setTodos(todos =>
            // map through each todo item, if id of item matches the id of item being update, then replace with updated value, else keep value the same
            todos.map(todo =>
                todo.id === id ? { ...todo, todoTask: updatedTodoTask } : todo
            )
        )
    }

    return (
        <div>
            <NewTodoForm addTodo={addTodo} />
            {todos.map(({ id, todoTask }) =>
                <Todo
                    key={id}
                    id={id}
                    todoTask={todoTask}
                    removeTodo={removeTodo}
                    updateTodo={update}
                />
            )}
        </div>
    )
}

export default TodoList
