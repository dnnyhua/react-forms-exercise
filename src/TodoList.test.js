import { render, fireEvent, getByText } from "@testing-library/react";
import TodoList from "./TodoList";

// function addTodo(todoList, todoTask = "testing") {
//     const todoInput = todoList.getByLabelText("Todo");
//     fireEvent.change(todoInput, { target: { value: todoTask } });
//     const submitButton = todoList.getBytext("Add");
//     fireEvent.click(submitButton)
// }


function addTodo(todoList, task = "write tests") {
    const taskInput = todoList.getByLabelText("Todo");
    fireEvent.change(taskInput, { target: { value: task } });
    const submitButton = todoList.getByText("Add");
    fireEvent.click(submitButton);
}

it("renders without crashing", function () {
    render(<TodoList />);
});


it("matches snapshot", function () {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});


it("can add a new todo item", function () {
    const { queryByText, getByLabelText } = render(<TodoList />);
    const input = getByLabelText("Todo");
    const addBtn = queryByText("Add");
    const editBtn = queryByText("Edit");

    expect(input).toHaveValue("")
    expect(queryByText('Walk the dog')).toBeInTheDocument();
    expect(queryByText('buy Chocolate Milk')).not.toBeInTheDocument();

    fireEvent.change(input, { target: { value: 'buy Chocolate Milk' } });
    fireEvent.click(addBtn);

    expect(queryByText('buy Chocolate Milk')).toBeInTheDocument();
    expect(editBtn).toBeInTheDocument();
    // expect(getByText("X")).toBeInTheDocument();
});