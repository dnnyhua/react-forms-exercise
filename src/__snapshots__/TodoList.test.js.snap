// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form
      class="NewTodoForm"
    >
      <label
        for="todoTask"
      >
        Todo
      </label>
      <input
        id="todoTask"
        name="todoTask"
        placeholder="Add Todo Item"
        type="text"
        value=""
      />
      <button>
        Add
      </button>
    </form>
    <div
      class="Todo"
    >
      <li>
        Walk the dog
      </li>
      <button>
        Edit
      </button>
      <button>
        X
      </button>
    </div>
  </div>
</DocumentFragment>
`;
